USER-PACE pair_style integration into LAMMPS

yury-lysogorskiy opened the pull request:
**Summary**

Performant implementation of atomic cluster expansion and interface to LAMMPS

**Related Issue(s)**

**Author(s)**

Copyright 2021 (c) Yury Lysogorskiy^1, Cas van der Oord^2, Anton Bochkarev^1,
Sarath Menon^1, Matteo Rinaldi^1, Thomas Hammerschmidt^1, Matous Mrovec^1,
Aidan Thompson^3, Gabor Csanyi^2, Christoph Ortner^4, Ralf Drautz^1

^1: Ruhr-University Bochum, Bochum, Germany
^2: University of Cambridge, Cambridge, United Kingdom
^3: Sandia National Laboratories, Albuquerque, New Mexico, USA
^4: University of British Columbia, Vancouver, BC, Canada

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

yury-lysogorskiy commented:
@akohlmey I started to prototype the USER-PACE integration in this PR. I have following questions:
- src/USER-PACE/Install.sh: what should this file do? Patch some Makefiles?
- lib/pace/Install.py: shold this script download the USER-PACE source from our repo? Where from it should be called? Automatically or by user?

akohlmey commented:
> @akohlmey I started to prototype the USER-PACE integration in this PR. I have following questions:
> 
> * src/USER-PACE/Install.sh: what should this file do? Patch some Makefiles?

that file serves two purposes:
- copy/update/remove files from the src/PACKAGE folder to the src/ folder when using the conventional build system
- update src/Makefile.package and src/Makefile.package.setting to include the variables and settings from the lib/package/Makefile.lammps file so that the library and include paths and flags from it are properly include (or removed) from the conventional build process when the package is installed or uninstalled.

> * lib/pace/Install.py: shold this script download the USER-PACE source from our repo? Where from it should be called? Automatically or by user?

It can be called directly or via 'make lib-package args="..."'
in the conventional build system, this is a separate step. 

Please see the descriptions of the conventional build system in general for packages with libraries in lib and also the examples I mentioned previously for further information and templates. The individual packages should have README files in the lib/package folder with some more instructions. Also individual Install.py files should have a --help option that will explain the available flags and their usage.

yury-lysogorskiy commented:
@akohlmey I did the modifications,  could, you take a look, please?
@athomps Could you, please, check the examples/USER/pace and doc/src, please?

athomps commented:
I was not able to compile with either cmake or make.  I was able to build `libpace.a`. `Makefile.lammps` in lib/pace seems to be missing.

yury-lysogorskiy commented:
> I was not able to compile with either cmake or make. I was able to build `libpace.a`. `Makefile.lammps` in lib/pace seems to be missing.

Sorry, forgot to add Makefile.lammps.

@athomps  How do you compile with cmake ?

athomps commented:
make works now.  I still could not figure out how to make cmake work.  The examples in examples/USER/pace are good.  It would be good to rename `potentials/Cu-PBE-core-rep.ace` to something more descriptive like `Cu_npjCompMat2021.ace` and a header line e.g. 
# DATE: 2021-04-07 UNITS: metal CONTRIBUTOR: Yury Lysogorskiy <yury.lysogorskiy@ruhr-uni-bochum.de> CITATION: ... (2021)

athomps commented:
Need to add a section on `USER-PACE` to `Build_package.rst`, but we to get cmake working first.

akohlmey commented:
> Need to add a section on `USER-PACE` to `Build_package.rst`, but we to get cmake working first.

I can have a look at the CMake stuff now, while I am waiting for the final merges for the next patch release to clear automated testing.

akohlmey commented:
@yury-lysogorskiy the archive at https://github.com/ICAMS/lammps-user-pace/archive/refs/tags/v.2021.2.3.tar.gz is broken as it contains the pair_pace.cpp and pair_pace.h sources.

akohlmey commented:
@yury-lysogorskiy also the conventional build procedure is broken since it *requires* CMake to build the downloaded library. The whole point of the conventional build procedure is to *not* require CMake.

akohlmey commented:
> @yury-lysogorskiy also the conventional build procedure is broken since it _requires_ CMake to build the downloaded library. The whole point of the conventional build procedure is to _not_ require CMake.

Just as a reminder, systems like RHEL 7.x or CentOS 7.x ship with CMake 2.x. CMake 3.x is available only via EPEL or manual installation.

akohlmey reviewed (changes requested):
Please see my comments for some requests for "formal" changes.

I have updated the code in src/USER-PACE/ to be aligned with the LAMMPS coding conventions and also to reflect recent changes in the core of LAMMPS to avoid certain "unsafe" practices and increase code reuse and convenience.
I also added a couple of unit test reference data files.

akohlmey commented:
@yury-lysogorskiy FYI the USER-PACE package cannot be compiled with the PGI compilers because several of the downloaded files are missing a newline at the end of the file. Example:
```
"/home/akohlmey/compile/lammps/build-pace/lammps-user-pace-v.2021.2.3/USER-PACE
          /ace_types.h", line 45: error: last line of file ends without a
          newline
  #endif
```
Also there are the following warnings (those also show up with clang++):
```
"/home/akohlmey/compile/lammps/build-pace/lammps-user-pace-v.2021.2.3/USER-PACE
          /ace_c_basisfunction.h", line 198: warning: function
          "ACEAbstractBasisFunction::_copy_from(const ACEAbstractBasisFunction
          &)" is hidden by "ACECTildeBasisFunction::_copy_from" -- virtual
          function override intended?
      void _copy_from(const ACECTildeBasisFunction &other) {

```

yury-lysogorskiy commented:
> @yury-lysogorskiy the archive at https://github.com/ICAMS/lammps-user-pace/archive/refs/tags/v.2021.2.3.tar.gz is broken as it contains the pair_pace.cpp and pair_pace.h sources.

the pair_pace.* files are necessary there, because this repository is used for `conda` package compilation and manual distribution. I hope the trick with filtering out the *pair_pace.\** files in `lib/pace/Makefile` and in `cmake/Modules/Packages/USER-PACE.cmake` would be acceptable.

yury-lysogorskiy commented:
> @yury-lysogorskiy also the conventional build procedure is broken since it _requires_ CMake to build the downloaded library. The whole point of the conventional build procedure is to _not_ require CMake.

I modified the compilation process in `lib/pace` to use  **make** instead of **CMake** and update `lib/pace/Install.py` correspondingly

yury-lysogorskiy commented on src/USER-PACE/pair_pace.cpp:
```diff
@@ -0,0 +1,458 @@
+^4: University of British Columbia, Vancouver, BC, Canada
+
+
+    This FILENAME is free software: you can redistribute it and/or modify
```
license terms are removed, only authorship remains

yury-lysogorskiy commented on src/USER-PACE/LICENSE:
```diff
@@ -0,0 +1,674 @@
+                    GNU GENERAL PUBLIC LICENSE
```
file removed

yury-lysogorskiy commented:
> @yury-lysogorskiy FYI the USER-PACE package cannot be compiled with the PGI compilers because several of the downloaded files are missing a newline at the end of the file. Example:
> 
> ```
> "/home/akohlmey/compile/lammps/build-pace/lammps-user-pace-v.2021.2.3/USER-PACE
>           /ace_types.h", line 45: error: last line of file ends without a
>           newline
>   #endif
> ```
> 
> Also there are the following warnings (those also show up with clang++):
> 
> ```
> "/home/akohlmey/compile/lammps/build-pace/lammps-user-pace-v.2021.2.3/USER-PACE
>           /ace_c_basisfunction.h", line 198: warning: function
>           "ACEAbstractBasisFunction::_copy_from(const ACEAbstractBasisFunction
>           &)" is hidden by "ACECTildeBasisFunction::_copy_from" -- virtual
>           function override intended?
>       void _copy_from(const ACECTildeBasisFunction &other) {
> ```

I added extra newlines at the ends of source/headers and removed "virtual" keyword in ACEAbstractBasisFunction::_copy_from. I hope that should remove the warnings thrown by clang++/PGI compilers

akohlmey commented:
> > @yury-lysogorskiy the archive at https://github.com/ICAMS/lammps-user-pace/archive/refs/tags/v.2021.2.3.tar.gz is broken as it contains the pair_pace.cpp and pair_pace.h sources.
> 
> the pair_pace.* files are necessary there, because this repository is used for `conda` package compilation and manual distribution. I hope the trick with filtering out the _pair_pace.*_ files in `lib/pace/Makefile` and in `cmake/Modules/Packages/USER-PACE.cmake` would be acceptable.

It is not a clean solution, but acceptable. The resulting headaches of having two different and possibly diverging sources for the same file are going to be yours and not ours.

akohlmey commented:
> I added extra newlines at the ends of source/headers and removed "virtual" keyword in ACEAbstractBasisFunction::_copy_from. I hope that should remove the warnings thrown by clang++/PGI compilers

There are a whole lot more (avoidable) warnings, but since they are in the external sources, we don't pay attention to them. For LAMMPS we try to keep (justified) compiler warnings to the minimum, as that increases the visibility of those that need addressing. a few cannot be avoided and some are better not silenced, since they make debugging with valgrind harder. I only mentioned this specific one, as it is often an indication of a bug in the code.

yury-lysogorskiy commented:
> > > @yury-lysogorskiy the archive at https://github.com/ICAMS/lammps-user-pace/archive/refs/tags/v.2021.2.3.tar.gz is broken as it contains the pair_pace.cpp and pair_pace.h sources.
> > 
> > 
> > the pair_pace.* files are necessary there, because this repository is used for `conda` package compilation and manual distribution. I hope the trick with filtering out the _pair_pace.*_ files in `lib/pace/Makefile` and in `cmake/Modules/Packages/USER-PACE.cmake` would be acceptable.
> 
> It is not a clean solution, but acceptable. The resulting headaches of having two different and possibly diverging sources for the same file are going to be yours and not ours.

Ok, very good, thank you. These files `pair_pace.h/cpp` from targ.gz and from `lammps/src/USER-PACE` are actually identical. So I think it is better to have duplicated identical files rahter than two different repos...

akohlmey commented:
@yury-lysogorskiy as far as the C++ source code is concerned, this seems to be working well enough for inclusion.
The remaining required action items are to finish the documentation and to figure out whether to distribute the potential file with the distribution or move it to download.lammps.org like we do with other large potential files. The provided file is not quite as large as those but larger than any other potential file currently in the distribution.

I have a couple more observations.
- how do you make certain, that the correct units are used? we usually embed the required units and other information into the potential files as a comment and add a check for that. But it is also possible to just add a check in the source code since the provided file seems to require "metal" units and if that is the only option
- it would be nice to have a short README file in the src/USER-PACE file, similar to what is - for example - in src/LATTE/README. Just some minimal info for people well versed with configuring/compiling LAMMPS

yury-lysogorskiy commented:
>  But it is also possible to just add a check in the source code since the provided file seems to require "metal" units and if that is the only option

we silently assume "metal" units (eV, Ang.). How could one check that in pair_pace.cpp? Which LAMMPS API to use?

akohlmey commented:
> > But it is also possible to just add a check in the source code since the provided file seems to require "metal" units and if that is the only option
> 
> we silently assume "metal" units (eV, Ang.). How could one check that in pair_pace.cpp? Which LAMMPS API to use?

something like this:
```
  // QUIP potentials are parameterized in metal units

  if (strcmp("metal",update->unit_style) != 0)
    error->all(FLERR,"QUIP potentials require 'metal' units");
```

athomps commented:
@yury-lysogorskiy I will test build again now and then add build instructions to `Packages_details.rst`

athomps commented on potentials/Cu-PBE-core-rep.ace:
```diff
@@ -0,0 +1,8980 @@
+nelements=1
```
It will be the 45th largest file in potentials dir, increasing dir size by about 1%, so I don't think we should exclude it.  There is some benefit to letting people take a look at it without having to build the package.

athomps commented:
Doc pages are complete

athomps reviewed (approved):
I wholeheartedly approve this PR.

athomps commented:
I tried to add this header line to `Cu-PBE-core-rep.ace`:

`# DATE: 2021-04-08 UNITS: metal CONTRIBUTOR: Yury Lysogorskiy <yury.lysogorskiy@ruhr-uni-bochum.de> CITATION: npj Comp. Mat., submitted (2021)
`
and got:

```  
terminate called after throwing an instance of 'std::invalid_argument'
what():  File 'Cu-PBE-core-rep.ace': couldn't read 'nelements', read:'#'. Expected format: 'nelements=[number]'
```

It would be nice if there was a way for the PACE library to accommodate that.
